### DISCUSSION TITLE
Use ContentTypeMigrationContext AddNewContentType to Add a new ContentType before Migration

### DISCUSSION THREAD

**Discussion by ProNotion** [Jumoo/uSyncMigrations#237] [Q&A]
I'm trying to work out if/how it might be possible to add a new ContentType before a migration and Change the ContentType of existing elements within a NestedContent property. I'm not sure if I am missing an existing way to do this by overthinking the solution but as it stands I've had to do the following:

1. Create my own NestedContentMigrator that inherits from `NestedContentMigrator` 
2. In the constructor I create the required content types if they do not exist using the `ContentTypeService`
3. Override the `GetConfigValues` method in which I have a switch statement to allow me to return an amended configuration value depending on the datatype or use the `base.GetConfigValues(dataTypeProperty, context)` method if there is no match. In the amended config I am updating the old content type alias for the new one created in step 1.

```
public override object? GetConfigValues(SyncMigrationDataTypeProperty dataTypeProperty,
	SyncMigrationContext context)
{
	if (string.IsNullOrWhiteSpace(dataTypeProperty.ConfigAsString))
		return base.GetConfigValues(dataTypeProperty, context);

	// If this is the slideshow slide editor, change the content type alias to the new doctype
	switch (dataTypeProperty.DataTypeAlias)
	{
		case "Slideshow Slide Editor":
			return UpdateSlideShowContentBlockConfigValues(dataTypeProperty, context);
		case "Content Block Editor":
			return UpdateContentBlockConfigValues(dataTypeProperty, context);
	}

	// Return the result from the base method
	return base.GetConfigValues(dataTypeProperty, context);
}

private object? UpdateSlideShowContentBlockConfigValues(SyncMigrationDataTypeProperty dataTypeProperty,
	SyncMigrationContext context)
{
	const string oldDoctypeAlias = "SlideShowSlide";
	const string newDoctypeAlias = "slideShowContentBlockSlide";

	var config = JsonConvert.DeserializeObject<NestedContentConfiguration>(dataTypeProperty.ConfigAsString);

	if (config == null)
	{
		return base.GetConfigValues(dataTypeProperty, context);
	}

	if (config.ContentTypes != null && config.ContentTypes.Any(c => c.Alias == oldDoctypeAlias))
	{
		config.ContentTypes[Array.FindIndex(config.ContentTypes, c => c.Alias == oldDoctypeAlias)].Alias =
			newDoctypeAlias;
	}

	return config;
}
```

4. I also override the `GetContentValue` method allowing me to update the content type of each row of nested content using a switch statement similar to step 3.

```
public override string? GetContentValue(SyncMigrationContentProperty contentProperty, SyncMigrationContext context)
{
	if (string.IsNullOrWhiteSpace(contentProperty.Value)) return string.Empty;

	var rowValues = JsonConvert.DeserializeObject<IList<NestedContentRowValue>>(contentProperty.Value);
	if (rowValues == null) return string.Empty;

	foreach (var row in rowValues)
	{
		if (row.Id == default)
		{
			row.Id = Guid.NewGuid();
		}

		// Update the content type to the new content type alias (which is an element type)
		switch (row.ContentTypeAlias)
		{
			case "SlideShowSlide":
				row.ContentTypeAlias = "slideShowContentBlockSlide";
				break;
			case "boxSliderSection":
				row.ContentTypeAlias = "boxSliderSectionContentBlock";
				break;
		}

		foreach (var property in row.RawPropertyValues)
		{
			var editorAlias =
				context.ContentTypes.GetEditorAliasByTypeAndProperty(row.ContentTypeAlias, property.Key);
			if (editorAlias == null) continue;

			try
			{
				var migrator = context.Migrators.TryGetMigrator(editorAlias.OriginalEditorAlias);
				if (migrator != null)
				{
					row.RawPropertyValues[property.Key] = migrator.GetContentValue(
						new SyncMigrationContentProperty(
							row.ContentTypeAlias, property.Key, row.ContentTypeAlias, property.Value?.ToString()),
						context);
				}
			}
			catch (Exception ex)
			{
				throw new Exception(
					$"Nested Error: [{editorAlias.OriginalEditorAlias} -{property.Key}] : {ex.Message}", ex);
			}
		}
	}

	return JsonConvert.SerializeObject(rowValues, Formatting.Indented);
}
```

It works but I wrote it before discovering AddNewContentType in ContentTypeMigrationContext but am unsure if it would be possible and better to somehow use that instead.

Any thoughts @KevinJump ?

**Comment by KevinJump** [comment]
Hi,

Yeah you could use the AddNewContentType, - in the GetConfigValues method (as example in the archetype migrator https://github.com/Jumoo/uSyncMigrations/blob/6ea75b295e185a975cb7e06e8a02333638430fb8/uSync.Migrations.Migrators/Community/Archetype/ArchetypeToBlockListMigrator.cs#L60-L91 - this might be more complex than you need).

So in the order things run, 

1) the datatype stuff is ran (including GetConfig) this adds newContentType to a list
2) the migration of the content types runs, and also creates and new ContentTypes.
3) the content is migrated so the GetContentValue methods are calls .

usually, you won't need the content type before step 3 so adding it in the GetConfigValue method should be, ok?

In theory you can also have property-based migrators, so you can define the migrator you want to use based on the property alias or the contenttype_property alias. 

you would define that on your custom "migration plan" 

e.g 

https://github.com/Jumoo/uSyncMigrations/blob/6ea75b295e185a975cb7e06e8a02333638430fb8/MyMigrations/MyMigrationProfile.cs#L55-L62

I am not sure at the moment there might not be away to do this by datatypeAlias - maybe that should be added. 

that way you might still need a custom migrator, but you wouldn't have to worry with the base values (not a big saving really - might not actually give you much) - it would mean you could have two migrators here one for each type. 

-

**Reply by ProNotion** [reply]
Great, thanks for the quick reply @KevinJump - I'll work through this and see how I get on.
